src/snowflake/cli/api/metrics.py

```diff
@@ -54,17 +67,125 @@ class CLICounterField:
     PACKAGE_SCRIPTS = f"{_TypePrefix.FEATURES}.{_DomainPrefix.APP}.package_scripts"
 
 
+class _CLIMetricsStep:
+    """
+    class for holding metrics step data and encapsulating related operations
+    """
+
+    _id_counter = count(start=1, step=1)
+
+    # keys for dict representation
+    ID_KEY = "id"
+    NAME_KEY = "name"
+    PARENT_KEY = "parent"
+    PARENT_ID_KEY = "parent_id"
+    START_TIME_KEY = "start_time"
+    EXECUTION_TIME_KEY = "execution_time"
+    ERROR_KEY = "error"
+
+    def __init__(self, name: str, parent: _CLIMetricsStep | None = None):
+        if not name:
+            raise CLIMetricsInvalidUsageError("step name must not be empty")
+
+        # to disambiguate steps with same name
+        self._step_id: int = next(self._id_counter)
+        self._name: str = name
+
+        self._parent: Optional[str] = parent.name if parent is not None else None
+        self._parent_id: Optional[int] = parent.step_id if parent is not None else None
+
+        self._monotonic_start: Optional[
+            float
+        ] = None  # monotonic time for execution time
+        self._start_time: Optional[float] = None  # timestamp
+        self._execution_time: Optional[float] = None
+
+        self._error: Optional[str] = None
+
+    def start(self) -> None:
+        if self._monotonic_start:
+            raise CLIMetricsInvalidUsageError("step has already started")
+        self._start_time = time.time()
+        self._monotonic_start = time.monotonic()
+
+    def end(self, error: Optional[BaseException] = None) -> None:
+        if self._execution_time:
+            raise CLIMetricsInvalidUsageError("step has already ended")
+        if not self._monotonic_start:
+            raise CLIMetricsInvalidUsageError("step has not started")
+
+        if error:
+            self._error = type(error).__name__
+
+        self._execution_time = time.monotonic() - self._monotonic_start
+
+    @property
+    def step_id(self) -> int:
+        return self._step_id
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def parent(self) -> Optional[str]:
+        return self._parent
+
+    @property
+    def parent_id(self) -> Optional[int]:
+        return self._parent_id
+
+    @property
+    def start_time(self) -> float:
+        if not self._start_time:
+            raise CLIMetricsInvalidUsageError("start_time accessed before step started")
+        return self._start_time
+
+    @property
+    def execution_time(self) -> float:
+        if not self._execution_time:
+            raise CLIMetricsInvalidUsageError(
+                "execution_time accessed before step ended"
+            )
+        return self._execution_time
+
+    @property
+    def error(self) -> Optional[str]:
+        """
+        name of the error class that was raised during the execution of this step
+        """
+        return self._error
+
+    def to_dict(self) -> Dict[str, Union[str, int, float, None]]:
```

Comment:
this is for reporting? I'd add a comment to that effect, since it makes this method a contract that should be stable.